airbyte-ci/connectors/base_images/base_images/common.py

```diff
@@ -0,0 +1,172 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import inspect
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import Path
+from typing import final
+
+import dagger
+import semver
+from base_images import consts
+
+
+@dataclass
+class PlatformAwareDockerImage:
+    image_name: str
+    tag: str
+    sha: str
+    platform: dagger.Platform
+
+    def get_full_image_name(self) -> str:
+        return f"{self.image_name}:{self.tag}@sha256:{self.sha}"
+
+
+class BaseImageVersionError(ValueError):
+    """Raised when the version is not in the expected format."""
+
+    pass
+
+
+class SanityCheckError(Exception):
+    """Raised when a sanity check fails."""
+
+    pass
+
+
+class PlatformAvailabilityError(ValueError):
+    """Raised when the platform is not supported by an image."""
+
+    pass
+
+
+class BaseBaseImage(Enum):
+    pass
+
+
+class AirbyteConnectorBaseImage(ABC):
+    """An abstract class that represents an Airbyte base image.
+    Please do not declare any Dagger with_exec instruction in this class as in the abstract class context we have no guarantee about the underlying system used in the base image.
+    """
+
+    @property
+    @abstractmethod
+    def base_base_image(cls) -> BaseBaseImage:
```

Comment:
This is not a typo: we're talking about the base image of the base image in this context.